async_tcp_server.cpp

```diff
@@ -0,0 +1,101 @@
+#include "async_tcp_server.hpp"
+#include <algorithm>
+#include <cstring>
+#include <iostream>
+#include <sys/select.h>
+
+AsyncTCPServer::AsyncTCPServer() {}
+
+AsyncTCPServer::AsyncTCPServer(Socket_sptr master_socket, fd_set fdsetRead): m_fdsetRead(fdsetRead)
+{
+    FD_ZERO(&m_fdsetRead);
+    FD_SET(master_socket->GetSocketHandle(), &m_fdsetRead);
+    m_aSockets.push_back(master_socket);
+}
+
+Socket_sptr AsyncTCPServer::GetSocketByFD(int fd) const
+{
+    for(auto socket : m_aSockets)
+    {
+        if (socket->GetSocketHandle() == fd)
+            return socket;
+    }
+
+    return Socket_sptr(nullptr);
+}
+
+AsyncTCPServer::~AsyncTCPServer()
+{
+    try
+    {
+        GetMasterSocket()->Close();
+    }
+    catch (std::runtime_error & re)
+    {
+        std::cout << "~AsyncTCPServer() failed" << std::endl;
+    }
+}
+
+Socket_vector AsyncTCPServer::Select(int nfds, timeval* timeout)
+{
+    fd_set temp_set;
+    FD_ZERO(&temp_set);
+    memcpy(&temp_set, &m_fdsetRead, sizeof(m_fdsetRead));
+    int ready_sockets_count = select(nfds, &temp_set,  NULL, NULL, timeout);
+    if(ready_sockets_count < 0)
+        throw std::runtime_error("select() failed");
+
+    Socket_vector ready_sockets;
+    ready_sockets.clear();
+
+    for(int i = 0; i < nfds && ready_sockets_count > 0; ++i)
+    {
+        Socket_sptr socket = GetSocketByFD(i);
+        if(FD_ISSET(i, &temp_set) && socket.get())
+        {
+            ready_sockets.push_back(socket);
+            ready_sockets_count--;
+        }
+    }
+
+    return ready_sockets;
+}
+
+Socket_sptr AsyncTCPServer::GetMasterSocket() const
```

Comment:
Lets imagine next scenario:
1. We're creating `AsyncTCPServer` - it is created with master socket inside.
2. We're accepting few connections - they're all added to vector.
3. Now we'll call `AsyncTCPServer::CloseConnection` for master socket - it will be removed from vector.
4. Next call of `GetMasterSocket` function will absolutely legally return first socket from vector, which is obviously not the master socket.
Think about it ;)